### PR TITLE
More Node bindings

### PR DESCRIPTION
**Pull request by jack-signal**
Aes256GcmSiv
ProtocolAddress
PreKeyRecord
PreKeyBundle
PreKeySignalMessage
SignalMessage

### CODE REVIEW

File: node/index.ts [hunk 1]
```diff
@@ -75,3 +135,229 @@ export class PrivateKey {
...
+    signed_prekey_id: number,
+    signed_prekey: PublicKey,
+    signed_prekey_signature: Buffer,
+    identity_key: PublicKey
+  ) {
+    this.nativeHandle = SC.PreKeyBundle_New(
+      registration_id,
+      device_id,
+      prekey_id,
+      prekey != null ? prekey._unsafeGetNativeHandle() : null,
```
Comment (jrose-signal) [inline node/index.ts:156]: TypeScript has optional chaining like Swift, so you can make this `prekey?._unsafeGetNativeHandle()`.

File: node/index.ts [hunk 2]
```diff
@@ -75,3 +135,229 @@ export class PrivateKey {
...
+
+  private constructor(handle: SignalClient.SignalMessage) {
+    this.nativeHandle = handle;
+  }
+
+  _unsafeGetNativeHandle(): SignalClient.SignalMessage {
+    return this.nativeHandle;
+  }
+
+  new(
```
Comment (jrose-signal) [inline node/index.ts:253]: Missing `static`.

File: node/index.ts [hunk 3]
```diff
@@ -75,3 +135,229 @@ export class PrivateKey {
...
+}
+
+export class PreKeySignalMessage {
+  private readonly nativeHandle: SignalClient.PreKeySignalMessage;
+
+  private constructor(handle: SignalClient.PreKeySignalMessage) {
+    this.nativeHandle = handle;
+  }
+
+  new(
```
Comment (jrose-signal) [inline node/index.ts:318]: Missing `static`

File: node/index.ts [hunk 4]
```diff
@@ -75,3 +135,229 @@ export class PrivateKey {
...
+        preKeyId,
+        signedPreKeyId,
+        baseKey._unsafeGetNativeHandle(),
+        identityKey._unsafeGetNativeHandle(),
+        signalMessage._unsafeGetNativeHandle()
+      )
+    );
+  }
+
+  deserialize(buffer: Buffer): PreKeySignalMessage {
```
Comment (jrose-signal) [inline node/index.ts:340]: Missing `static`

File: node/index.ts [hunk 5]
```diff
@@ -75,3 +135,229 @@ export class PrivateKey {
     );
   }
 }
+
+export class PreKeyBundle {
+  private readonly nativeHandle: SignalClient.PreKeyBundle;
+
+  constructor(
```
Comment (jrose-signal) [inline node/index.ts:142]: I feel weird having some classes use constructors and others use `static new()`. Maybe they should all use `new()`, with the constructor limited to the "from native handle" use case?

File: node/index.ts [hunk 6]
```diff
@@ -11,6 +11,58 @@ const SC = bindings('libsignal_client_' + os.platform()) as typeof SignalClient;
...
+    handle: SignalClient.ProtocolAddress
+  ): ProtocolAddress {
+    return new ProtocolAddress(handle);
+  }
+
+  static new(name: string, deviceId: number): ProtocolAddress {
+    return new ProtocolAddress(SC.ProtocolAddress_New(name, deviceId));
+  }
+
+  name(): string {
```
Comment (jrose-signal) [inline node/index.ts:57]: I think we already checked with the Desktop folks over whether to use method or property syntax for things like this, but it's worth double-checking.

File: node/index.ts (hunk 1 above)
Comment (jack-signal) [reply node/index.ts:156]: Unfortunately TypeScript maps null to undefined thus `prekey?._unsafeGetNativeHandle()` is `PublicKey | undefined` not `PublicKey | null` - I assume there is a way around this but haven't found it yet.

File: node/index.ts (hunk 6 above)
Comment (jack-signal) [reply node/index.ts:61]: Evan's comment (in #124) on this "I'd call this getPublicKey and keep it as a function. I rarely see accessors used in JavaScript-land. (This is anecdotal, and other Desktop devs might have different experiences.)"

**Review by jrose-signal — APPROVED** [review]

**Pull request merged**